src/Collection/Collection.php

```diff
@@ -73,17 +73,29 @@ public function unserialize($collection)
     }
 
     /**
-     * Throws an exception.
+     * {@inheritDoc}
      *
-     * Issuing a count on a Collection can have many side effects, some making the
-     * Collection unusable after the count operation.
-     *
-     * @return void
-     * @throws \LogicException
+     * @return int
      */
     public function count()
     {
-        throw new LogicException('You cannot issue a count on a Collection.');
+        $traversable = $this->optimizeUnwrap();
```

Comment:
Why don't you do the same operation as [ResultSetDecorator](https://github.com/cakephp/cakephp/blob/424f2eb1f292a64bd53dfc29e478dd995eebe38d/src/Datasource/ResultSetDecorator.php#L36-L43)?

Reply:
I know ResultSetDecorator In all of the cases is wrapping a countable object. Here I do not know that. But here I have the functions to dive down to an array.

I could change the decorator to be similar to this.

Reply:
But if the innter iterator is a Countable, why don't you believe the return value of count()? I think the problematic iterator is a FilterIterator here. FilterIterator would need to call iterator_count() or need to count array. But I think other Collection classes can call Countable::count().